### ISSUE TITLE
Test all of the file-related twisted.web.wsgi code-paths with each of the possible file-like objects they might be used with

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4114]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4114|
|Type|enhancement|
|Created|2009-11-16 21:01:55Z|
|Branch|https://github.com/twisted/twisted/tree/wsgi.input-4114-3|

Twisted Web sometimes uses a `cStringIO.StringIO`, sometimes a `tempfile.TemporaryFile`, for `request.content`.  These types are subtly incompatible.  The Twisted Web WSGI tests should exercise the relevant code against each of these types (and perhaps be generally factored to be amenable to testing against new types in case the need for such arises).  This will catch problems like the one described in [#4029](https://github.com/twisted/twisted/issues/4029).

<details><summary>Searchable metadata</summary>

```
trac-id__4114 4114
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_wsgi_input_4114_3 branches/wsgi.input-4114-3
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1258405315000000 1258405315000000
changetime__1260077508000000 1260077508000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [27513]) Branching to 'wsgi.input-4114'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [27519]) Branching to 'wsgi.input-4114-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [27518]) Merge wsgi-none-readline-4029

Author: rikyu
Reviewer: exarkun
Fixes: [#4029](https://github.com/twisted/twisted/issues/4029)
Refs: [#4114](https://github.com/twisted/twisted/issues/4114)

Add a test for WSGI requests which end up with a `tempfile.TemporaryFile`
for `Request.content` and fix the implementation to deal with this case
by not passing `None` for the length argument to `readline` (which temp
files do not accept, unlike StringIOs).  WSGI applications can now receive
requests with large bodies.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [27545]) Branching to 'wsgi.input-4114-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Okay, I think that's pretty thoroughly tested.  A couple more potential issues revealed, although the WSGI spec is unclear on exactly which of these cases should be supported.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

A couple of pyflakes:
```
twisted/web/test/test_wsgi.py:15: redefinition of unused 'tempfile' from line 10
twisted/web/test/test_wsgi.py:476: redefinition of unused 'version' from line 29
twisted/web/test/test_wsgi.py:524: redefinition of unused 'http' from line 27
```

+1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [27682]) Fix some pyflakes

refs [#4114](https://github.com/twisted/twisted/issues/4114)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [27683]) Merge wsgi.input-4114-3

Author: exarkun
Reviewer: therve
Fixes: [#4114](https://github.com/twisted/twisted/issues/4114)

Split up the existing tests for wsgi.input into many separate test methods
and apply the tests based on several different underlying low-level file-like
objects.  Also fix a few inconsistencies which were pointed out by this more
exhaustive testing.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
